server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -411,83 +416,130 @@ public void initializeSnapshot(SnapshotId snapshotId, List<IndexId> indices, Met
                                     snapshotId, index, finalShardId), ex);
                             }
                         }
                     }
-                }
+                    if (outstandingIndices.decrementAndGet() == 0) {
+                        deleteUnreferencedIndices(repositoryData, updatedRepositoryData, listener);
+                    }
+                });
             }
+        } else {
+            deleteUnreferencedIndices(repositoryData, updatedRepositoryData, listener);
+        }
+    }
 
-            // cleanup indices that are no longer part of the repository
-            final Collection<IndexId> indicesToCleanUp = Sets.newHashSet(repositoryData.getIndices().values());
-            indicesToCleanUp.removeAll(updatedRepositoryData.getIndices().values());
-            final BlobContainer indicesBlobContainer = blobStore().blobContainer(basePath().add("indices"));
-            for (final IndexId indexId : indicesToCleanUp) {
+    private void deleteUnreferencedIndices(RepositoryData repositoryData, RepositoryData updatedRepositoryData,
+            ActionListener<Void> listener) {
+        // cleanup indices that are no longer part of the repository
+        final Collection<IndexId> indicesToCleanUp = Sets.newHashSet(repositoryData.getIndices().values());
+        indicesToCleanUp.removeAll(updatedRepositoryData.getIndices().values());
+        final BlobContainer indicesBlobContainer = blobStore().blobContainer(basePath().add("indices"));
+        if (indicesToCleanUp.isEmpty()) {
+            // We're done, no indices to clean up
+            listener.onResponse(null);
+            return;
+        }
+        final AtomicInteger outstanding = new AtomicInteger(indicesToCleanUp.size());
+        for (final IndexId indexId : indicesToCleanUp) {
+            threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(() -> {
                 try {
                     indicesBlobContainer.deleteBlob(indexId.getId());
                 } catch (DirectoryNotEmptyException dnee) {
                     // if the directory isn't empty for some reason, it will fail to clean up;
                     // we'll ignore that and accept that cleanup didn't fully succeed.
                     // since we are using UUIDs for path names, this won't be an issue for
                     // snapshotting indices of the same name
-                    logger.debug(() -> new ParameterizedMessage("[{}] index [{}] no longer part of any snapshots in the repository, " +
+                    logger.error(() -> new ParameterizedMessage("[{}] index [{}] no longer part of any snapshots in the repository, " +
                         "but failed to clean up its index folder due to the directory not being empty.", metadata.name(), indexId), dnee);
-                } catch (IOException ioe) {
+                } catch (Exception e) {
```

Comment:
revert this change?

Reply:
This was a conscious choice, I'd actually rather log all exceptions here directly + it simplifies things since we don't need to use the ActionRunnable here if we catch everything. I adjusted the logic a little though and invoked the grouped listener in a `finally` to make things a little clearer (and avoid the potential odd block from a future assertion trippin somewhere).